### PR TITLE
#3 payment handlers

### PR DESCRIPTION
**Pull request by lojzatran**
Issue: https://github.com/commercetools/commercetools-paypal-plus-integration/issues/3

Part of this task is also to edit the config in Travis so the tests won't fail anymore.

### CODE REVIEW

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        if (isAnyPropertyNull(projectKey, clientId, clientSecret,
+                ppPClientId, ppPClientSecret, ppPClientMode)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new TenantConfig(projectKey, clientId, clientSecret,
+                ppPClientId, ppPClientSecret, ppPClientMode));
+    }
+
+    private boolean isAnyPropertyNull(String... properties) {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: `org.apache.commons.lang3.StringUtils#isAnyBlank(CharSequence... css)`
`org.apache.commons.lang3.StringUtils#isNoneBlank(CharSequence... css)`

(the dependency included transparently by `commercetools-models`)

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        String projectKey = env.getProperty(tenantName + ".ctp.client.projectKey");
+        String clientId = env.getProperty(tenantName + ".ctp.client.clientId");
+        String clientSecret = env.getProperty(tenantName + ".ctp.client.clientSecret");
+        String ppPClientId = env.getProperty(tenantName + ".paypalPlus.client.clientId");
+        String ppPClientSecret = env.getProperty(tenantName + ".paypalPlus.client.clientSecret");
+        String ppPClientMode = env.getProperty(tenantName + ".paypalPlus.client.mode");
+
+        if (isAnyPropertyNull(projectKey, clientId, clientSecret,
+                ppPClientId, ppPClientSecret, ppPClientMode)) {
+            return Optional.empty();
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: we have a problem here (and not only): so far we silently create empty objects, so in the future we will have a lot of troubles debugging the issues why the payments are note processed!

We have couple of choices:
  - Add at least logging of _warn_ or _error_ if tenant specific handlers can't be created with explicit reasons why
  - add the error message to response body, like "tenant %s can't be created because of missing parameters XXX"
  - make a custom exception thrown if some tenant configs or tenant factories, or tenant handlers can't be created

Basically, we could aggregate all this solutions to some common solution, but we should not stick with silence.

File: src/main/java/com/commercetools/pspadapter/util/ClientConfigurationUtils.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import io.sphere.sdk.http.AsyncHttpClientAdapter;
+import io.sphere.sdk.http.HttpClient;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.DefaultAsyncHttpClient;
+import org.asynchttpclient.DefaultAsyncHttpClientConfig;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.TimeUnit;
+
+public class ClientConfigurationUtils {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/util/ClientConfigurationUtils.java]: `Sphere` or `Ctp` `ClientConfigurationUtils`

File: src/test/java/com/commercetools/service/main/PaymentHandlerProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest(classes = Application.class)
+public class PaymentHandlerProviderTest {
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/service/main/PaymentHandlerProviderTest.java]: no cases with missing tenant config

File: src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java [hunk 5]
```diff
@@ -46,10 +55,12 @@ public static int cleanupPaymentTable(SphereClient sphereClient) {
...
      * @return number of read/deleted items
      */
-    public static <EntityType> int cleanupTable(
+    private static <EntityType> int cleanupTable(
             @Nonnull final SphereClient client,
             @Nonnull final Supplier<SphereRequest<PagedQueryResult<EntityType>>> querySupplier,
-            @Nonnull final Function<EntityType, SphereRequest<EntityType>> deleteFunction) {
+            @Nonnull final Function<EntityType, SphereRequest<EntityType>> deleteFunction,
+            @Nonnull String resourceName) {
+        logger.debug("Cleanup " + resourceName + " table");
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java]: `logger.debug("Cleanup {} table", resourceName);`

see https://www.slf4j.org/manual.html#typical_usage
and  https://www.slf4j.org/faq.html#logging_performance

File: src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java [hunk 6]
```diff
@@ -46,10 +55,12 @@ public static int cleanupPaymentTable(SphereClient sphereClient) {
      * @param <EntityType>   type of items to read/delete
      * @return number of read/deleted items
      */
-    public static <EntityType> int cleanupTable(
+    private static <EntityType> int cleanupTable(
             @Nonnull final SphereClient client,
             @Nonnull final Supplier<SphereRequest<PagedQueryResult<EntityType>>> querySupplier,
-            @Nonnull final Function<EntityType, SphereRequest<EntityType>> deleteFunction) {
+            @Nonnull final Function<EntityType, SphereRequest<EntityType>> deleteFunction,
+            @Nonnull String resourceName) {
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java]: 👍

File: src/test/java/com/commercetools/testUtil/customTestConfigs/CtpClientConfig.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import io.sphere.sdk.client.SphereClientConfig;
+import io.sphere.sdk.client.SphereClientFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class CtpClientConfig {
+
+    @Value("${paypalplus-integration-test.ctp.client.projectKey}")
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/testUtil/customTestConfigs/CtpClientConfig.java]: since this will be environment variable names - never use dash (**-**) in the configuration keys: only alpha-numeric, dot (**.**) and underscore (**_**)

**Comment by andrii-kovalenko-ct** [comment]
Also, on application start mandatory verify, and (maybe) pre-cache all the tenants config values:

 - [ ] parse `tenants` config list
 - [ ] parse all the `<tenantName>.ctp.*` and `<teantName>.paypalPlus.*` values and verify they are not blank (neither null, not white-space string)
 - [ ] (optionally) create  `TenantConfig` for each _tenatName_ and on every call return the cached value.

Guess `TenantConfigFactory#new()` is a proper candidate for all this features. We should throw some custom exception with reasonable message if some tenant configs can't be created (like _missing some config values for tenant [%s]_), and the whole application should not be run in this case!

File: build.gradle [hunk 8]
```diff
@@ -49,6 +49,7 @@ dependencies {
     }
 
     testCompile "org.assertj:assertj-core:${depVersions.assertJ}"
+    testCompile group: 'org.mockito', name: 'mockito-core', version: '2.8.47'
```
Comment (lojzatran) [inline build.gradle]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/bff73319538dd2c675f29cd1dabd91c12751b23e

File: src/main/java/com/commercetools/payment/handler/CommercetoolsHandlePaymentsController.java [hunk 9]
```diff
@@ -1,29 +1,41 @@
...
 public class CommercetoolsHandlePaymentsController {
 
     private static final String template = "Hello, payment [%s]!";
 
     private final StringTrimmerEditor stringTrimmerEditor;
+    private final PaymentHandlerProvider paymentHandlerProvider;
 
     @Autowired
-    public CommercetoolsHandlePaymentsController(StringTrimmerEditor stringTrimmerEditor) {
+    public CommercetoolsHandlePaymentsController(StringTrimmerEditor stringTrimmerEditor,
```
Comment (lojzatran) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsHandlePaymentsController.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/019083ba3bc0d653dfd79023de1f7309ac2556ec

File: src/main/java/com/commercetools/payment/handler/CommercetoolsHandlePaymentsController.java [hunk 10]
```diff
@@ -1,29 +1,41 @@
...
     }
 
     @RequestMapping(
             method = RequestMethod.GET,
             value = "/{tenantName}/commercetools/handle/payments/{paymentId}")
     public PaymentDemo handlePayments(@PathVariable String tenantName,
                                       @PathVariable String paymentId) {
-
+        Optional<PaymentHandler> paymentHandlerOpt = paymentHandlerProvider.getPaymentHandler(tenantName);
+        paymentHandlerOpt.ifPresent(paymentHandler -> paymentHandler.handlePayment(paymentId));
```
Comment (lojzatran) [inline src/main/java/com/commercetools/payment/handler/CommercetoolsHandlePaymentsController.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/22cac3ddbe83b725c391847e77e801701da7a12b

File: src/main/java/com/commercetools/pspadapter/executor/CtpExecutor.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import com.commercetools.service.ctp.CartService;
+import com.commercetools.service.ctp.OrderService;
+import com.commercetools.service.ctp.PaymentService;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A wrapper class for all services that communicate with CT platform
+ */
+public class CtpExecutor {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/executor/CtpExecutor.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/5f474cec6c1d122578a3f2ec2329b258f6b0c524

File: src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutor.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package com.commercetools.pspadapter.executor;
+
+import com.commercetools.service.paypalPlus.PaypalPlusPaymentService;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A wrapper class for all services that communicate with Paypal Plus
+ */
+public class PaypalPlusExecutor {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutor.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/5f474cec6c1d122578a3f2ec2329b258f6b0c524

File: src/main/java/com/commercetools/pspadapter/executor/CtpExecutorFactory.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    private final TenantConfigFactory config;
+
+    @Autowired
+    public CtpExecutorFactory(@Nonnull TenantConfigFactory config) {
+        this.config = config;
+    }
+
+    public Optional<CtpExecutor> getCtpExecutor(@Nonnull String tenantName) {
+        Optional<TenantConfig> tenantConfigOpt = this.config.getTenantConfig(tenantName);
+        if (tenantConfigOpt.isPresent()) {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/executor/CtpExecutorFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/99545991df8c3f3ed86de0b0eadd10815b1f8540

File: src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutorFactory.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    private final TenantConfigFactory config;
+
+    @Autowired
+    public PaypalPlusExecutorFactory(@Nonnull TenantConfigFactory config) {
+        this.config = config;
+    }
+
+    public Optional<PaypalPlusExecutor> getPaypalPlusExecutor(@Nonnull String tenantName) {
+        Optional<TenantConfig> tenantConfigOpt = config.getTenantConfig(tenantName);
+        if (tenantConfigOpt.isPresent()) {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutorFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/99545991df8c3f3ed86de0b0eadd10815b1f8540

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    private final CtpExecutor ctpExecutor;
+    private final PaypalPlusExecutor paypalPlusExecutor;
+
+    public PaymentHandler(@Nonnull CtpExecutor ctpExecutor,
+                          @Nonnull PaypalPlusExecutor paypalPlusExecutor) {
+        this.ctpExecutor = ctpExecutor;
+        this.paypalPlusExecutor = paypalPlusExecutor;
+    }
+
+    public void handlePayment(@Nonnull String paymentId){
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/22cac3ddbe83b725c391847e77e801701da7a12b

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+        ctpExecutor.getCartService().getByPaymentId(paymentId);
+        // mapping
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/22cac3ddbe83b725c391847e77e801701da7a12b#diff-99689329c2a9889010715255d37afc27R25

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java (hunk 2 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/8720a23ab3207db6407eb032711ca6a54ca2c4d9

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java (hunk 1 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/840d38c1b404269afb824353090780e91802a861

File: src/main/java/com/commercetools/pspadapter/util/ClientConfigurationUtils.java (hunk 3 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/util/ClientConfigurationUtils.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/2adbc8516d8754545687ea9bb07c72c1cea7fe5c

File: src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java (hunk 5 above)
Comment (lojzatran) [inline src/test/java/com/commercetools/testUtil/ctpUtil/CleanupTableUtil.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/e8f4034b93c67c74c8386e94aba02ea0a0d82567

**Comment by lojzatran** [comment]
I validated the configs here: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/8720a23ab3207db6407eb032711ca6a54ca2c4d9

File: src/test/java/com/commercetools/testUtil/customTestConfigs/CtpClientConfig.java (hunk 7 above)
Comment (lojzatran) [inline src/test/java/com/commercetools/testUtil/customTestConfigs/CtpClientConfig.java]: Problem solved - I removed this file.

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java (hunk 2 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/01c47bf3fd9cab52f5de5268e125c09ae3111609

**Comment by lojzatran** [comment]
I add some tests: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/ad938e415284b5248d97b9af6354904da9de3398

File: src/main/java/com/commercetools/pspadapter/executor/CtpExecutorFactory.java (hunk 13 above)
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/executor/CtpExecutorFactory.java]: `.flatMap(Optional.of(x))` -> `map(x)`

File: src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutorFactory.java (hunk 14 above)
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutorFactory.java]: `flatMap(Optional.of(x))` => `map(x)`

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java (hunk 2 above)
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: same here about logging: log that config for {tenant name} is invalid.

File: src/main/java/com/commercetools/pspadapter/tenant/TenantProperties.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+@Configuration
+@ConfigurationProperties("tenantConfig")
+@Validated
+public class TenantProperties {
+
+    @Valid
+    private List<Tenant> tenants;
+
+    public List<Tenant> getTenants() {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantProperties.java]: shy not a HashMap? than you do:
```java
tenantProperties.getTenants().stream()
    .filter(t -> tenantName.equals(t.getName()))
    .findAny();
```
what is super inefficient in this case!

Basically, you should not expose the tenants collection at all, add method: `Optional<Tenant> TenantProperties#getTenantByName(String name)`

File: src/test/java/com/commercetools/testUtil/mockObjects/TenantPropertiesMock.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
+package com.commercetools.testUtil.mockObjects;
+
+import com.commercetools.pspadapter.tenant.TenantProperties;
+
+import java.util.Collections;
+
+public class TenantPropertiesMock {
```
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/testUtil/mockObjects/TenantPropertiesMock.java]: i have not found any <i><b>multy</b></i>tenancy tests

File: src/main/java/com/commercetools/pspadapter/tenant/TenantProperties.java (hunk 17 above)
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantProperties.java]: also keep the `tenants` as `Map<String, Tenant>`

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java (hunk 2 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: I don't want to put logging there because it would make multiple logging if tenant name is invalid. Maybe we should log only from 1 place (e.g. from controllers only)?

File: src/main/java/com/commercetools/pspadapter/tenant/TenantProperties.java [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+import org.springframework.validation.annotation.Validated;
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+@Configuration
+@ConfigurationProperties("tenantConfig")
+@Validated
+public class TenantProperties {
```
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantProperties.java]: nice approach 👍

File: src/test/java/com/commercetools/testUtil/customTestConfigs/CtpClientConfig.java (hunk 7 above)
Comment (andrii-kovalenko-ct) [inline src/test/java/com/commercetools/testUtil/customTestConfigs/CtpClientConfig.java]: no class - no problem. though it might be incompatible with previous projects configuration, like `payone-integration-service`. Follow [the discussion](https://github.com/commercetools/commercetools-paypal-plus-integration/issues/3#issuecomment-318323604)

File: src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java (hunk 2 above)
Comment (andrii-kovalenko-ct) [inline src/main/java/com/commercetools/pspadapter/tenant/TenantConfigFactory.java]: ok, fine

File: src/main/java/com/commercetools/pspadapter/executor/CtpExecutorFactory.java (hunk 13 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/executor/CtpExecutorFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/32c52e0121d8cad97b0f51171a7fafda415bff6c

File: src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutorFactory.java (hunk 14 above)
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/executor/PaypalPlusExecutorFactory.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/32c52e0121d8cad97b0f51171a7fafda415bff6c

File: src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImpl.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Autowired
+    public PaymentHandlerProviderImpl(@Nonnull CtpExecutorFactory ctpFactory,
+                                      @Nonnull PaypalPlusExecutorFactory pPPFactory) {
+        this.ctpFactory = ctpFactory;
+        this.pPPFactory = pPPFactory;
+    }
+
+    @Override
+    public Optional<PaymentHandler> getPaymentHandler(@Nonnull String tenantName) {
```
Comment (lojzatran) [inline src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandlerProviderImpl.java]: https://github.com/commercetools/commercetools-paypal-plus-integration/pull/23/commits/345575eeb6653039b530653cf220c68cfb004bef
